src/madness/mra/bc.h

```diff
@@ -0,0 +1,188 @@
+/*
+  This file is part of MADNESS.
+
+  Copyright (C) 2007,2010 Oak Ridge National Laboratory
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+
+  For more information please contact:
+
+  Robert J. Harrison
+  Oak Ridge National Laboratory
+  One Bethel Valley Road
+  P.O. Box 2008, MS-6367
+
+  email: harrisonrj@ornl.gov
+  tel:   865-241-3937
+  fax:   865-572-0680
+*/
+
+#ifndef MADNESS_MRA_BC_H__INCLUDED
+#define MADNESS_MRA_BC_H__INCLUDED
+
+
+/// \file bc.h
+/// \brief Provides BoundaryConditions
+/// \ingroup mrabcext
+
+#include <madness/world/madness_exception.h>
+
+#include <array>
+#include <cstddef>
+#include <iostream>
+
+namespace madness {
+
+enum BCType {
+  BC_ZERO,
+  BC_PERIODIC,
+  BC_FREE,
+  BC_DIRICHLET,
+  BC_ZERONEUMANN,
+  BC_NEUMANN
+};
+
+/*!
+  \brief This class is used to specify boundary conditions for all operators
+  \ingroup mrabcext
+
+  Exterior boundary conditions (i.e., on the simulation domain)
+  are associated with operators (not functions).  The types of
+  boundary conditions available are in the enum BCType.
+
+  The default boundary conditions are obtained from the FunctionDefaults.
+  For non-zero Dirichlet and Neumann conditions additional information
+  must be provided when derivative operators are constructed. For integral
+  operators, only periodic and free space are supported.
+*/
+template <std::size_t NDIM> class BoundaryConditions {
+private:
+  // Used to use STL vector but static data on  a MAC was
+  // causing problems.
+  BCType bc[NDIM * 2];
+
+public:
+  /// Constructor. Default boundary condition set to free space
+  BoundaryConditions(BCType code = BC_FREE) {
+    for (std::size_t i = 0; i < NDIM * 2; ++i)
```

Comment:
Why not use `std::fill` or `std::fill_n` here? Same question for the assignment operator. Not sure if I'm missing something.